File: 217-Contains Duplicate/contains_duplicate.dart
main() {
  final nums = [2, 3, 2];
  ContainsDuplicate.containsDuplicate(nums);
}

abstract class ContainsDuplicate {
  static bool containsDuplicate(List<int> nums) {
    //
    // First attempt 02 July 2024 11:09
    // Runtime: 400ms, Memory: 180.35MB
    // Time complexity: O(n), Space complexity: O(n)
    // HashSet
    //
    final occurrenceMap = <int>{};

    for (int i = 0; i < nums.length; i++) {
      if (occurrenceMap.contains(nums[i])) {
        return true;
      }

      occurrenceMap.add(nums[i]);
    }

    return false;
  }
}

File: 217-Contains Duplicate/ContainsDuplicate.kt
fun main() {
    val nums = intArrayOf(1, 2, 3, 1)

    println("Output: ${ContainsDuplicate.containsDuplicate(nums)}")
}

object ContainsDuplicate {
    fun containsDuplicate(nums: IntArray): Boolean {
        //
        // First attempt 02 July 2024 16:54
        // Runtime: 245ms, Memory: 41.80MB
        // Time complexity: O(n), Space complexity: O(1)
        // HashSet
        //
        val occurrenceMap = hashSetOf<Int>()

        for (num in nums) {
            if (occurrenceMap.contains(num)) return true
            occurrenceMap.add(num)
        }

        return false
    }
}